db/silkworm/etl/file_provider.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+   Copyright 2020 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "file_provider.hpp"
+
+#include <boost/filesystem/operations.hpp>
+#include <silkworm/etl/error.hpp>
+
+namespace silkworm::etl {
+
+    namespace fs = boost::filesystem;
+
+FileProvider::FileProvider(std::string& working_path, int id) : id_{id} {
+    fs::path path{fs::path(working_path) / fs::path("tmp-" + std::to_string(id))};
+    filename_ = path.string();
+    file_.open(filename_, std::ios_base::in | std::ios_base::out | std::ios_base::trunc);
+    if (!file_.is_open()) throw etl_error(strerror(errno));
+}
+
+void FileProvider::flush(Buffer &buffer) {
+    head_t head{};
+
+    // Verify we have enough space to store all data
+    auto &entries{buffer.get_entries()};
+    size_t data_size{buffer.size() + entries.size() * sizeof(head_t)};
+    fs::path workdir(fs::path(filename_).parent_path());
+    if (fs::space(workdir).available < data_size) {
+        throw etl_error("Insufficient disk space");
+    }
+
+    for (const auto &entry : entries) {
+        head.lengths[0] = entry.key.size();
+        head.lengths[1] = entry.value.size();
+        file_.write((const char *)head.bytes, 8);
+        file_.write((const char *)entry.key.data(), entry.key.size());
+        file_.write((const char *)entry.value.data(), entry.value.size());
+        if (file_.fail()) {
+            throw etl_error(strerror(errno));
+        }
+    }
+    file_.seekg(0);
+}
+
+std::optional<std::pair<db::Entry, int>> FileProvider::read_entry() {
+    head_t head{};
+    file_.read((char *)head.bytes, 8);
+    if (file_.eof()) {
+        return std::nullopt;
+    }
+    if (file_.fail()) {
+        throw etl_error(strerror(errno));
+    }
+    auto key{new unsigned char[head.lengths[0]]};
```

Comment:
There is a memory leak here: memory is allocated with `new`, but is nowhere released with `delete`.
One possible solution to avoid the leak: don't use `ByteView`, which doesn't own the memory, use `Bytes` instead. Something like

In a header file:
```
namespace silkworm::etl {

struct Entry {
    Bytes key;
    Bytes value;
    size_t size() const noexcept { return key.size() + value.size(); }
};
```

In `file_provider.cpp`:
```
// Entry from etl, not db namespace
std::optional<std::pair<Entry, int>> FileProvider::read_entry() {
  ...
  Entry entry;
  entry.key.resize(head.lengths[0]);
  entry.value.resize(head.lengths[1]);
  file_.read((char *)entry.key.data(), entry.key.size());
  file_.read((char *)entry.value.data(), entry.value.size());
  ...
```

Reply:
Name overlapping (even though in different namespaces) is a bit frustrating.
I think I'll make it a public struct within Fileprovider class